benchmark/evaluation/compare_near_eq_and_exact.py

```diff
@@ -0,0 +1,292 @@
+    if (stdev_log_rho_pi(np.array(Ws)) > threshold):
+        message = "stdev_log_rho_pi(Ws) > threshold\n({:.3f} > {:.3f})".format(stdev_log_rho_pi(np.array(Ws)), threshold)
+        warnings.warn(message, RuntimeWarning)
+
+
+    return np.array(Ws)
+
+
+def sample_from_rho(noneq_sim, n_steps=1000):
+    # (x0, v0) drawn from pi
+    x0 = noneq_sim.sample_x_from_equilibrium()
+    v0 = noneq_sim.sample_v_given_x(x0)
+
+    # (x, v) drawn from rho
+    W_shad_forward = noneq_sim.accumulate_shadow_work(x0, v0, n_steps)["W_shad"]
+    x = get_state_as_mdtraj(noneq_sim.simulation)
+    v = noneq_sim.simulation.context.getState(getVelocities=True).getVelocities(asNumpy=True)
+
+    return {'x': x,
+            'v': v,
+            'W_shad_forward': W_shad_forward,
+            }
+
+
+def outer_sample_naive(index=0, noneq_sim=None, marginal="full", n_inner_samples=100, n_steps=1000):
+    rho_sample = sample_from_rho(noneq_sim, n_steps)
+    x, v, W_shad_forward = rho_sample['x'], rho_sample['v'], rho_sample['W_shad_forward']
+
+    Ws = collect_inner_samples_naive(x, v, noneq_sim, marginal, n_steps, n_inner_samples)
+
+    return {
+        "xv": (x, v),
+        "Ws": Ws,
+        "estimate": estimate_from_work_samples(Ws),
+        "W_shad_forward": W_shad_forward
+    }
+
+
+def outer_sample_adaptive(noneq_sim=None, marginal="full", n_steps=1000, initial_size=100, batch_size=5, threshold=0.1,
+                          max_samples=1000):
+    rho_sample = sample_from_rho(noneq_sim, n_steps)
+    x, v, W_shad_forward = rho_sample['x'], rho_sample['v'], rho_sample['W_shad_forward']
+
+    Ws = collect_inner_samples_until_threshold(x, v, noneq_sim, marginal, initial_size, batch_size, n_steps, threshold,
+                                               max_samples)
+
+    return {
+        "xv": (x, v),
+        "Ws": Ws,
+        "estimate": estimate_from_work_samples(Ws),
+        "W_shad_forward": W_shad_forward
+    }
+
+
+def noneq_sim_factory(testsystem_name, scheme, dt, collision_rate):
+    testsystem = testsystems[testsystem_name]
+    integrator = LangevinSplittingIntegrator(splittings[scheme],
+                                             temperature=temperature,
+                                             collision_rate=collision_rate,
+                                             timestep=dt * unit.femtosecond)
+    noneq_sim = NonequilibriumSimulator(testsystem, integrator)
+    return noneq_sim
+
+
+def process_outer_samples(outer_samples):
+    """Compute near-equilibrium and "exact" estimates of D_KL from work samples"""
+
+    # estimate D_KL as a sample average over x ~ rho of log(rho(x) / pi(x))
+    new_estimate = np.mean([s["estimate"] for s in outer_samples], 0)
+
+    # near-equilibrium estimate
+    W_F_hat = np.mean([s["W_shad_forward"] for s in outer_samples])
+    W_R_hat = np.mean([s["Ws"][0] for s in outer_samples])  # picking the first W_R sample associated with each W_F
```

Comment:
It's not obvious to me that this is the work accumulated over the second half of the trajectory (after velocity randomization). How can I figure out what this is doing?

Reply:
`outer_samples` is a list of dicts, and each element's `Ws` item is a list of inner-loop work samples from the second half of the trajectory, and I pick an arbitrary sample from that. I could have replaced `[0]` in L183 with `[np.random.randint(len(s['Ws']))]`, for example